### PR TITLE
Support STATUS message

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Switch to the new STATUS message instead of HELLO. 
Support sending STATUS prior to a genesis event.
Avoid storing a reference to the Netty bytebuf messages are originally parsed from in the Peer status data.

### CODE REVIEW

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/PeerManager.java [hunk 1]
```diff
@@ -111,6 +112,11 @@ public Peer getPeer(Connection conn) {
     return connectedPeerMap.get(conn.getSecureSession().getRemoteId());
   }
 
+  public Optional<Peer> getPeer(PeerId peerId) {
+    final Peer peer = connectedPeerMap.get(peerId);
+    return peer != null && peer.hasReceivedHello() ? Optional.of(peer) : Optional.empty();
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/PeerManager.java]: (discussion) Since we're filtering based on `hasReceivedHello `, should this method be renamed to something like `getPeerIfAvailable()`?  Would we ever want to retrieve a peer before the `Status` request has completed?

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/Peer.java [hunk 2]
```diff
@@ -61,4 +69,69 @@ public boolean isInitiator() {
       final RpcMethod<I, O> method, I request) {
     return rpcMethods.invoke(method, connection, request);
   }
+
+  public boolean hasReceivedHello() {
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/Peer.java]:
```suggestion
  public boolean hasSentStatus() {
```

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/Peer.java [hunk 3]
```diff
@@ -37,8 +41,12 @@ public Peer(Connection connection, RpcMethods rpcMethods) {
     this.rpcMethods = rpcMethods;
   }
 
-  public void receivedHelloMessage(final HelloMessage message) {
-    remoteHello.complete(message);
+  public void receivedStatusMessage(final StatusMessage message) {
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/Peer.java]: (nit / optional):
```suggestion
  public void processStatusMessage(final StatusMessage message) {
```

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/PeerManager.java [hunk 4]
```diff
@@ -129,6 +135,6 @@ private void onDisconnectedPeer(Peer peer) {
   }
 
   public int getPeerCount() {
-    return connectedPeerMap.size();
+    return (int) connectedPeerMap.values().stream().filter(Peer::hasReceivedHello).count();
```
Comment (mbaxter) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/PeerManager.java]: (discussion) Since we're filtering our peers, should this be renamed to `getAvailablePeerCount()` or similar?

File: networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/JvmLibP2pSmokeIntegrationTest.java [hunk 5]
```diff
@@ -45,9 +48,25 @@ public void shouldConnectToPeers() throws Exception {
...
+    assertThat(network2ViewOfPeer1.getStatus().getHeadRoot()).isEqualTo(Bytes32.ZERO);
+    assertThat(network2ViewOfPeer1.getStatus().getHeadSlot()).isEqualTo(UnsignedLong.ZERO);
+
+    final Peer network1ViewOfPeer2 =
+        network1.getPeerManager().getPeer(network2.getPeerId()).orElseThrow();
+    assertThat(network1ViewOfPeer2.getStatus().getForkVersion()).isEqualTo(Fork.VERSION_ZERO);
+    assertThat(network1ViewOfPeer2.getStatus().getFinalizedRoot()).isEqualTo(Bytes32.ZERO);
+    assertThat(network1ViewOfPeer2.getStatus().getFinalizedEpoch()).isEqualTo(UnsignedLong.ZERO);
+    assertThat(network1ViewOfPeer2.getStatus().getHeadRoot()).isEqualTo(Bytes32.ZERO);
+    assertThat(network1ViewOfPeer2.getStatus().getHeadSlot()).isEqualTo(UnsignedLong.ZERO);
```
Comment (mbaxter) [inline networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/JvmLibP2pSmokeIntegrationTest.java:69]: Should we add a test with non-zero values?

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/networking/libp2p/rpc/StatusMessage.java [hunk 6]
```diff
@@ -24,15 +24,15 @@
 import tech.pegasys.artemis.util.SSZTypes.SSZContainer;
 import tech.pegasys.artemis.util.sos.SimpleOffsetSerializable;
 
-public final class HelloMessage implements SimpleOffsetSerializable, SSZContainer {
+public final class StatusMessage implements SimpleOffsetSerializable, SSZContainer {
 
   private final Bytes4 forkVersion;
```
Comment (jrhea) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/networking/libp2p/rpc/StatusMessage.java]: you might want to consider renaming this headForkVersion to be consistent with the spec

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/networking/libp2p/rpc/StatusMessage.java [hunk 7]
```diff
@@ -75,36 +75,36 @@ public boolean equals(Object obj) {
...
-        && Objects.equals(this.headSlot(), other.headSlot());
+            this.getForkVersion().getWrappedBytes(), other.getForkVersion().getWrappedBytes())
+        && Objects.equals(this.getFinalizedRoot(), other.getFinalizedRoot())
+        && Objects.equals(this.getFinalizedEpoch(), other.getFinalizedEpoch())
+        && Objects.equals(this.getHeadRoot(), other.getHeadRoot())
+        && Objects.equals(this.getHeadSlot(), other.getHeadSlot());
   }
 
-  public Bytes4 forkVersion() {
+  public Bytes4 getForkVersion() {
```
Comment (jrhea) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/networking/libp2p/rpc/StatusMessage.java]: rename getHeadForkVersion

**Review by mbaxter — APPROVED** [review]

**Review by jrhea — APPROVED** [review]
A couple of suggestions; otherwise, lgtm

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/PeerManager.java (hunk 1 above)
Comment (ajsutton) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/PeerManager.java]: Makes sense to me.

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/Peer.java (hunk 2 above)
Comment (ajsutton) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/Peer.java]: Rejigged this area a bit.  This is indicative that we've received status data from the remote peer.  We can receive later updates though so I've swapped the underlying field from a `CompletableFuture` to a `volatile Optional` to support that.

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/Peer.java (hunk 3 above)
Comment (ajsutton) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/Peer.java]: Because `updateStatus`.

File: networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/JvmLibP2pSmokeIntegrationTest.java (hunk 5 above)
Comment (ajsutton) [reply networking/p2p/src/integration-test/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/JvmLibP2pSmokeIntegrationTest.java:69]: That's the next PR - it would currently fail due to a couple of bugs in our Resq/Resp framework.  We are holding  references to a Netty ByteBuf without retain so it get's released underneath us and we're actually always echoing back the request we receive, not the response we generated.
